### PR TITLE
fix: change multi-group dependency resolution

### PR DESCRIPTION
**Pull request by MikeDombo**
**Issue #, if available:**

**Description of changes:**
Change the multi-group dependency resolution to re-resolve all group dependencies by asking the cloud instead of trying to lock to the active version. This change completely removes the INSTALLED_COMPONENT_NOT_FOUND root cause.

**Why is this change necessary:**

**How was this change tested:**
- [ ] Updated or added new unit tests.
- [ ] Updated or added new integration tests.
- [ ] Updated or added new end-to-end tests.
- [ ] If my code makes a remote network call, it was tested with a proxy.

**Any additional information or context required to review the change:**

**Documentation Checklist:**
 - [ ] Updated the README if applicable.

**Compatibility Checklist:**
- [ ] I confirm that the change is backwards compatible.
- [ ] Any modification or deletion of public interfaces does not impact other plugin components.
- [ ] For external library version updates, I have reviewed its change logs and Nucleus does not consume 
  any deprecated method or type.

Refer to [Compatibility Guidelines](/COMPATIBILITY.md) for more information.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/componentmanager/DependencyResolver.java [hunk 1]
```diff
@@ -89,49 +89,47 @@ public List<ComponentIdentifier> resolveDependencies(DeploymentDocument document
...
+        // resolve updated version from the cloud, update version requirement map
+        for (String targetComponent : combinedTargetComponents) {
+            resolveComponentDependencies(targetComponent, componentNameToVersionConstraints,
                     resolvedComponents, componentIncomingReferenceCount,
-                    (name, requirements) -> componentManager.resolveComponentVersion(name, requirements));
+                    (name, requirements) ->
+                            componentManager.resolveComponentVersion(name, requirements));
         }
+        logger.atInfo().setEventType("resolve-all-group-dependencies-end")
+                .log("Done resolving all groups dependencies");
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/DependencyResolver.java:124]: remove log. We log on L132

File: src/test/java/com/aws/greengrass/componentmanager/DependencyResolverTest.java [hunk 2]
```diff
@@ -597,6 +600,110 @@ void GIVEN_component_A_B2_WHEN_dependencies_conflict_THEN_throws_no_available_ve
...
+                        new DeploymentPackageConfiguration(componentD, true, v1_0_0.getValue())),
+                Collections.emptyList(), "mockGroup2", "mockGroup2", "mockGroup2",
+                1L, FailureHandlingPolicy.DO_NOTHING, componentUpdatePolicy, configurationValidationPolicy);
+
+        Map<String, Set<ComponentRequirementIdentifier>> otherGroupRootPackages = new HashMap<>();
+        Set<ComponentRequirementIdentifier> group1RootPackages = new HashSet<>();
+        group1RootPackages.add(new ComponentRequirementIdentifier(componentA, Requirement.buildNPM("1.0.0")));
+        otherGroupRootPackages.put("mockGroup1", group1RootPackages);
+
+        context.runOnPublishQueueAndWait(() -> System.out.println("Waiting for queue to finish updating the config"));
```
Comment (MikeDombo) [inline src/test/java/com/aws/greengrass/componentmanager/DependencyResolverTest.java:698]: no print, just do `context.waitForPublishQueueToClear`

File: src/main/java/com/aws/greengrass/componentmanager/DependencyResolver.java [hunk 3]
```diff
@@ -89,49 +89,47 @@ public List<ComponentIdentifier> resolveDependencies(DeploymentDocument document
...
+                            .log("Found component configuration");
+                    componentNameToVersionConstraints.putIfAbsent(e.getPackageName(), new HashMap<>());
+                    componentNameToVersionConstraints.get(e.getPackageName())
+                            .put(document.getGroupName(), Requirement.buildNPM(e.getResolvedVersion()));
+                    targetComponentsToResolve.add(e.getPackageName());
+                });
+
+        Set<String> combinedTargetComponents = new LinkedHashSet<>(targetComponentsToResolve);
+        combinedTargetComponents
+                .addAll(getOtherGroupsTargetComponents(otherGroupsToRootComponents, componentNameToVersionConstraints));
```
[the remark this answers is not in the document]
Comment (alter-mage) [reply src/main/java/com/aws/greengrass/componentmanager/DependencyResolver.java:109]: I think, just before we loop on combinedTargetComponents to resolveComponentDependencies, componentNameToVersionConstraints should have the requirements of root components of all groups (others + current). So I dont think it matters whether I add the current group requirements before other group requirements, they will resolved at the same time. Not sure if that's your concern..

[the remark this answers is not in the document]
Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/componentmanager/DependencyResolver.java:109]: OK this looks fine then.

**Review by junfuchen99 — APPROVED** [review]

**Review by indougnito — APPROVED** [review]

**Pull request merged**